[docs] -- Adds FCL API reference (#729)
onflow/flow-js-sdk: docs/api.md

## Methods

---

## Query and Mutate Flow with Cadence

If you want to run arbitrary Cadence scripts on the blockchain, these methods offer a convenient way to do so **without having to build, send, and decode interactions**.

## `fcl.query({...options})`

Allows you to submit scripts to query the blockchain.

### Options

_Pass in the following as a single object with the following keys.All keys are optional unless otherwise stated._

| Key       | Type                                    | Description                                                                                                 |
| --------- | --------------------------------------- | ----------------------------------------------------------------------------------------------------------- |
| `cadence` | string **(required)**                   | A valid cadence script.                                                                                     |
| `args`    | [ArgumentFunction](#`ArgumentFunction`) | Any arguments to the script if needed should be supplied via a function that returns an array of arguments. |

> | `limit`   | number                                  | Compute (Gas) limit for query. Got to the [documentation about computation cost](https://docs.onflow.org/flow-go-sdk/building-transactions/#gas-limit) for information about how computation cost is calculated on Flow.        |

Reviewer: ```suggestion
| `limit`   | number                                  | Compute (Gas) limit for query. Read the [documentation about computation cost](https://docs.onflow.org/flow-go-sdk/building-transactions/#gas-limit) for information about how computation cost is calculated on Flow.        |
```

> 

Reviewer: ```suggestion
| `limit`   | number                                  | Compute (Gas) limit for query. Read the [documentation about computation cost](https://docs.onflow.org/flow-go-sdk/building-transactions/#gas-limit) for information about how computation cost is calculated on Flow.         |

```

### Options

_Pass in the following as a single object with the following keys. All keys are optional unless otherwise stated._

| Key        | Type                                            | Description                                                                                                                             |
| ---------- | ----------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------- |
| `cadence`  | string **(required)**                           | A valid cadence transaction.                                                                                                            |
| `args`     | [ArgumentFunction](##`ArgumentFunction`)        | Any arguments to the script if needed should be supplied via a function that returns an array of arguments.                             |

> | `limit`    | number                                          | Compute limit for query. :tomato: WHAT UNITS ARE THESE IN?                                                                              |

Reviewer: ```suggestion
| `limit`    | number                                          | Compute (Gas) limit for query. Go to the [documentation about computation cost](https://docs.onflow.org/flow-go-sdk/building-transactions/#gas-limit) for information about how computation cost is calculated on Flow.                                                                               |
```

> | `limit`    | number                                          | Compute (Gas) limit for query. Go to the [documentation about computation cost](https://docs.onflow.org/flow-go-sdk/building-transactions/#gas-limit) for information about how computation cost is calculated on Flow.                                                                               |

Reviewer: ```suggestion
| `limit`    | number                                          | Compute (Gas) limit for query. Read the [documentation about computation cost](https://docs.onflow.org/flow-go-sdk/building-transactions/#gas-limit) for information about how computation cost is calculated on Flow.                                                                               |
```

## Methods

---

## Query and Mutate Flow with Cadence

If you want to run arbitrary Cadence scripts on the blockchain, these methods offer a convenient way to do so **without having to build, send, and decode interactions**.

## `fcl.query({...options})`

Allows you to submit scripts to query the blockchain.

### Options

_Pass in the following as a single object with the following keys.All keys are optional unless otherwise stated._

| Key       | Type                                    | Description                                                                                                 |
| --------- | --------------------------------------- | ----------------------------------------------------------------------------------------------------------- |
| `cadence` | string **(required)**                   | A valid cadence script.                                                                                     |
| `args`    | [ArgumentFunction](#`ArgumentFunction`) | Any arguments to the script if needed should be supplied via a function that returns an array of arguments. |

> | `limit`   | number                                  | Compute limit for query. :tomato: WHAT UNITS ARE THESE IN?                                                  |

Reviewer: ```suggestion
| `limit`   | number                                  | Compute limit for query. Got to the [documentation](https://docs.onflow.org/flow-go-sdk/building-transactions/#gas-limit) for information about how Gas is calculated on Flow.                                          |
```

Author: Can we put a reccomended value for testing -> Maybe the same number used in kitty-items?

Reviewer: This number is covered in the docs.

Reviewer 2: There is a spelling mistake in the suggested update: `Got to the [documentation]` -> `Go to the [documentation]`

### Options

_Pass in the following as a single object with the following keys. All keys are optional unless otherwise stated._

| Key        | Type                                            | Description                                                                                                                             |
| ---------- | ----------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------- |
| `cadence`  | string **(required)**                           | A valid cadence transaction.                                                                                                            |
| `args`     | [ArgumentFunction](##`ArgumentFunction`)        | Any arguments to the script if needed should be supplied via a function that returns an array of arguments.                             |

> | `limit`    | number                                          | Compute limit for query. :tomato: WHAT UNITS ARE THESE IN?                                                                              |

Reviewer: ```suggestion
| `limit`    | number                                          | Compute (Gas) limit for query. Got to the [documentation](https://docs.onflow.org/flow-go-sdk/building-transactions/#gas-limit) for information about how computation cost is calculated on Flow.                                                                              |
```

Reviewer 2: There is a spelling mistake in the suggested update: `Got to the [documentation]` -> `Go to the [documentation]`

---

## `CurrentUserObject`

| Key         | Value Type             | Default   | Description                                                                                                                                                       |
| ----------- | ---------------------- | --------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `addr`      | [Address](##`Address`) | `null`    | The public address of the current user                                                                                                                            |

> | `cid`       | string                 | `null`    | :tomato: TODO                                                                                                                                                     |

Reviewer: ```suggestion
| `cid`       | string                 | `null`    | The [content identifier](https://docs.ipfs.io/concepts/content-addressing/) for user metadata if provided by wallets.                                                                                                                                  |
```

> | `expiresAt` | number                 | `null`    | If the current session expires the value will be non-null. :tomato: UNSURE                                                                                        |

Reviewer: ```suggestion
| `expiresAt` | number                 | `null`    |   Allows wallets to specify a time-frame for a valid session.  |
```

| `f_type`    | string                 | `'USER'`  | A type identifier used internally by FCL.                                                                                                                         |

> | `f_vsn`     | string                 | `'1.0.0'` | The version to use when passing messages between the provider and the dapp. :tomato: UNSURE                                                                       |

Reviewer: ```suggestion
| `f_vsn`     | string                 | `'1.0.0'` | fcl protocol version.                                                                       |
```

| `loggedIn`  | boolean                | `null`    | If the user is logged in.                                                                                                                                         |

> | `services`  | [ServiceObject]        | `[]`      | A list of services that offer specific functionality (eg. authorization) from the wallet provider to the logged in user. :tomato: More documentation coming soon. |

Reviewer: ```suggestion
| `services`  | [ServiceObject]        | `[]`      |  A list of trusted services that express ways of interacting with the current user's identity, including means to further discovery, authentication, authorization, or other kinds interaction. |
```

Reviewer: @orodio Any examples of this in code we can link to?

## `CollectionObject`

A collection is a list of transactions that are contained in the same block.
| Key | Value Type | Description |
| ---- | ---------- | ----------- |
| `id` | string | The id of the collection. |
| `transactionIds` | [string] | The ids of the transactions included in the collection. |

## `ResponseObject`

The format of all responses in FCL returned from `fcl.send(...)`. For full details on the values and descriptions of the keys, view [here](https://github.com/onflow/flow-js-sdk/tree/master/packages/sdk/src/response).
| Key |
| ---- |
| `tag` |
| `transaction` |
| `transactionStatus` |
| `transactionId` |
| `encodedData` |
| `events` |
| `account` |
| `block` |
| `blockHeader` |
| `latestBlock` |
| `collection` |

## `Event Object`

| Key | Value Type | Description |
| ---- | ---------- | ----------- |
| `blockId` | string | ID of the block that contains the event. |
| `blockHeight` | number| Height of the block that contains the event. |
| `blockTimestamp` | string | The timestamp of when the block was sealed in a `DateString` format. eg. `'2021-06-25T13:42:04.227Z'` |
| `type` | [EventName](#EventName) | A string containing the event name. |

> | `transactionId` | object | :tomato: TODO |

Reviewer: ```suggestion
| `transactionId` | object | Can be used to query transaction information, eg. via a Flow block explorer. |
```

> | `transactionIndex` | object | :tomato: TODO |

Reviewer: ```suggestion
| `transactionIndex` | object | Used to prevent replay attacks. |
```

> | `eventIndex` | object | :tomato: TODO |

Reviewer: ```suggestion
| `eventIndex` | object | Used to prevent replay attacks. |
```

Author: For these 'eventIndex (number), transactionIndex (number), and transactionId (string)' fields, I think the type also needs to be changed from 'object' to 'string' or 'number' - my mistake but would you mind changing those as well

Reviewer: OK. Updated.

| `data` | any | The data emitted from the event. |

## `Transaction Statuses`

The status of a transaction will depend on the Flow blockchain network and which phase it is in as it completes and is finalized.
| Status Code | Description |
| ----------- | ----------- |
| `0` | Unknown
| `1` | Transaction Pending - Awaiting Finalization
| `2` | Transaction Finalized - Awaiting Execution
| `3` | Transaction Executed - Awaiting Sealing

> | `4` | Transaction Sealed - Transaction Complete

Reviewer: ```suggestion
| `4` | Transaction Sealed - Transaction Complete. At this point the transaction result has been committed to the blockchain.
```

| `5` | Transaction Expired

> :tomato: TODO better descriptions.

Reviewer: ```suggestion
```
This should be fine.
